### PR TITLE
Move account names out of KeyringController

### LINKED ISSUE #3946: Move account nicknames out of keyring controller
This is an old mistake, we should clean it up. `account-nicknames` probably deserve their own controller for now, would be a good candidate for cloud-syncing.

### PR DESCRIPTION
**Pull request by whymarrh**
Fixes #3946

This PR removes account nicknames from the `KeyringController` (see also [\[1\]][1]) and stores them in the `PreferencesController`.

I've kept the identities state (`.identities`) the same shape so as to reduce the amount of change.

  [1]:https://github.com/MetaMask/KeyringController/pull/13

### CODE REVIEW

**Comment by whymarrh** [comment]
This PR is blocked by MetaMask/KeyringController#16 and once that's merged I'll update the package.

**Comment by danfinlay** [comment]
I want to merge this very carefully, so it may not be today/ this sprint. Thanks for the reorg!

**Review by bitpshr — COMMENTED** [review]
Very clean PR syntax-wise, only a few very small nits. I'll leave it to someone with a bit more context to review this holistically.

File: app/scripts/controllers/preferences.js
Comment (whymarrh) [reply app/scripts/controllers/preferences.js]: It doesn't but the similar function in `KeyringController` used to be, so I kept it `async` for compatibility

Comment (bitpshr) [reply app/scripts/controllers/preferences.js]: Since you're already returning a Promise, I don't think marking this as `async` without using `await` statements within the function body actually does anything.

File: ui/app/components/pages/create-account/index.js
Comment (danfinlay) [inline ui/app/components/pages/create-account/index.js]: You probably could have saved yourself a lot of work by keeping the name `saveAccountLabel`, but this name is better, I just hope you didn't miss any instances!

**Review by danfinlay — COMMENTED** [review]
This looks generally good to me.

It cleans up the KeyringController, it decouples it from the PreferencesController, keeps code style very consistent, has lots of good tests, so pretty much good to go.

I would love if @tmashuang could QA this first, from a vault with account nicknames, just to manually make sure it migrates.

File: ui/app/components/pages/create-account/index.js
Comment (whymarrh) [reply ui/app/components/pages/create-account/index.js]: I think I got all the instances 😄 and agreed, I think this name is clearer

File: app/scripts/controllers/preferences.js
Comment (whymarrh) [reply app/scripts/controllers/preferences.js]: Good catch, though I'll undo this change as it's somewhat unrelated to this PR

**Comment by tmashuang** [comment]
First initial pass through was fine. When I simulated the how it will update, there was rendering issues. Which leads me to think we might need a migration for the account nicknames.

**Comment by whymarrh** [comment]
I have a migration that tries to do that but it could be wrong, how can I test that locally? (I've added unit test but that passes so it might need to be updated.)

File: app/scripts/controllers/address-book.js
Comment (whymarrh) [reply app/scripts/controllers/address-book.js]: Good idea, I'll update this

**Comment by danfinlay** [comment]
To test locally, I would
1. pull master
2. [build locally](https://github.com/MetaMask/metamask-extension#building-locally)
3. [add that build to chrome](https://github.com/MetaMask/metamask-extension/blob/master/docs/add-to-chrome.md)
4. set up an account
5. add a nickname
6. git pull this branch
7. Re-build on this branch (auto-rebuilds if you're using `npm start`)
8. After rebuild, open MetaMask again, enter password
9. see if nickname persisted
10. see if you can still update the nickname.

**Comment by whymarrh** [comment]
@danfinlay @tmashuang got the migration fixed if you want to take another look

**Comment by whymarrh** [comment]
I'm taking a look at the CI failures now that I've rebased this on `master`

**Comment by whymarrh** [comment]
@danfinlay @tmashuang this should be good for review again 😅

**Comment by whymarrh** [comment]
I've rebased this on the latest `develop` (cc @kumavis)

**Comment by whymarrh** [comment]
@danfinlay thanks for the detailed reproduction steps. I was initially confused because the shape of `identities` in the preferences controller after this PR is the same as it was in the keyring controller. It took me a bit to realize that, as you said, the migration is incorrect.

I've added a commit to fix this: 67310e15

```diff
@@ -34,7 +34,14 @@ function transformState (state) {
     return state
   }
 
-  state.PreferencesController.identities = state.KeyringController.walletNicknames
+  state.PreferencesController.identities = Object.keys(state.KeyringController.walletNicknames)
+    .reduce((identities, address) => {
+      identities[address] = {
+        name: state.KeyringController.walletNicknames[address],
+        address,
+      }
+      return identities
+    }, {})
   delete state.KeyringController.walletNicknames
   return state
 }
```

The unit test I had was incorrect and I seemingly didn't test the migration flow when I opened the PR originally.

> The way I found this issue is unfortunately tedious, and absolutely highlights an opportunity for stronger end-to-end tests of migrations in the UI:

We can certainly cover more cases with the end-to-end tests and that's what I'm working on next in #3949. In this case, I think the regular end-to-end tests would have covered this if they ran the migration (which I don't believe they do?). During CI, with the upcoming new UI e2e tests or even with the existing old UI e2e tests, if we ran the tests against `master`/`develop` (whatever the base of the PR is) before checking out the PR commits that should cover the migration process and I think that would've caught this.

**Comment by danfinlay** [comment]
Cool, I've opened a new issue for adding migrations to e2e tests here:
https://github.com/MetaMask/metamask-extension/issues/4283

**Comment by danfinlay** [comment]
@whymarrh Could you merge master into this, to get the security test fix?

**Review by kumavis — APPROVED** [review]

**Review by danfinlay — APPROVED** [review]
I've manually confirmed this is working! Just need to get that security test passing.

**Pull request merged**
